design.md

```diff
@@ -0,0 +1,306 @@
+8. The **parent** signals to **child** that it has been placed in a cgroup by closing pipe 2.
+9. The **parent** starts a goroutine calling `os/exec Cmd.Wait()`, waiting for the **child** to return in order to track `Job` status.
+10. Send started job details back to client.
+
+The **child** process executing `jobworker exec` will involve the following:
+1. Application starts, and the following file descriptors are opened:
+  - `/proc/self/fd/3` for the job details
+  - `/proc/self/fd/4` for the cgroup placement signaling
+2. Job details are read from from `/proc/self/fd/3` and used to create `os/exec Cmd` for the **grandchild** process.
+3. **child** waits for **parent** to signal that it has been placed in the correct cgroup on `/proc/self/fd/4`.
+4. **child** calls `os/exec Cmd.Start()` to start **grandchild**.
+5. **child** calls `os/exec Cmd.Wait()` to wait for **grandchild** to exit.
+6. **child** returns **grandchild's** exit code as its own.
+
+##### Stopping
+
+In order to stop a command, the following will be done:
+1. When the **parent** creates the **child** `os/exec Cmd` the **child** will have its own`pgid`.
+2. When the **child** creates the **grandchild** `os/exec Cmd` the `pgid` will be in the same `pgid` as the **child**.
+3. When the **parent** creates the **child** `os/exec Cmd` a context cancellation function will be stored as part of the `Job`. Any process seeking to stop a `Job` may call the cancellation function. This will kill both the **child** and **grandchild**.
+4. The Cgroup related to `Job` will also be removed in unrelated logic.
+
+##### Status
+
+The status of underlying **grandchild** will be returned through the **child** to **parent** as exitcodes. This status will be handling in step 9 of [Starting](#starting).
+
+###### JobStatus
+
+The JobStatus type indicates the status of a job. The status will be one of the following:
+
+- *running*: Job has been started and is currently running.
+- *stopped*: Job has been forcibly stopped by jobworker.
+- *exited*: Job has exited. (with exit code)
+
+##### Streaming Output
+
+Streaming output will involve reading a log file from `/var/log/jobworker` in chunks and writing it to the client until the client ends the stream.
+
+###### Implementation
+
+When an output stream is requested, the following will occur:
+1. `Job` will be fetched by `job.Service`
+2. `Job.OutputStream` will be called in its own goroutine and accept `context.Context` and `chan<- byte` arguments.
+3. `Job.OutputStream` will open the log file, and defer closing the fd.
+4. `Job.OutputStream` will start a goroutine listening on `<-ctx.Done()`.
+5. `Job.OutputStream` will read from log file, in chunks.
+6. When `Read` returns `io.EOF`, the process will sleep for a second and continue reading. See [Streaming Tradoffs](#streaming-tradeoffs) below.
+7. If the `context.Context` is cancelled, `job.OutputStream` will close the `io.ReaderCloser`, close the `chan<- byte`, and return.
```

Comment:
Let's assume the client never cancels, as they want to stream until the job is over. How do you detect/handle that?

Reply:
I think the simplest approach would be to check the job status. If the job is still `running`, keep reading. Otherwise, end the stream. I've added this to the document.